src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,171 @@
+        if (size == EMPTY_LIST) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index >= size && index < EMPTY_LIST) {
+            throw new IndexOutOfBoundsException("Index is out of list");
+        }
+
+        Node<T> newNode = new Node<>(value);
+
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> current = getNode(index);
+            newNode.prev = current.prev;
+            newNode.next = current;
+            if (current.prev != null) {
+                current.prev.next = newNode;
+            }
+            current.prev = newNode;
+            if (index == 0) {
+                head = newNode;
+            }
+        }
+
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null || list.isEmpty()) {
+            return;
+        }
+
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size && index < EMPTY_LIST) {
+            throw new IndexOutOfBoundsException("Index is out of list");
+        }
+
+        Node<T> current = getNode(index);
+        return current.item;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> current;
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index >= size && index < EMPTY_LIST) {
+            throw new IndexOutOfBoundsException("Index is out of list");
+        }
+
+        Node<T> current = getNode(index);
+        current.item = value;
+        return current.item;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size && index < EMPTY_LIST) {
+            throw new IndexOutOfBoundsException("Index is out of list");
+        }
+
+        Node<T> current = getNode(index);
+        unlinkNode(current);
+        size--;
+        return current.item;
+    }
+
+    private void unlinkNode(Node<T> node) {
+        node.item = null;
+        if (node.prev != null) {
+            node.prev.next = node.next;
+        } else {
+            head = node.next;
+        }
+        if (node.next != null) {
+            node.next.next = node.prev;
+        } else {
+            tail = node.prev;
+        }
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> current = getNode(object);
```

Comment:
You need to handle the case where `getNode(object)` may not find the node (if `object` is not in the list). Currently, it will always try to unlink the node, even if it's not found.